### ISSUE TITLE
re2ocaml can generate invalid escape sequences

### ISSUE THREAD

**Issue by eilvelia** [skvadrik/re2c#516]
```console
$ re2ocaml --utf8 -i src/re2c_lexer.ml.re2 -o src/re2c_lexer.ml
$ dune build
File "src/re2c_lexer.ml", line 633, characters 12-15:
633 |     | '\n'..'\f' ->
                  ^^^
Error: Illegal backslash escape in string or character (\f)
```

`\f` is not a valid escape in OCaml (https://ocaml.org/manual/5.2/lex.html#sss:character-literals), I think `\x0C` should be used instead.

This is generated roughly from
```
  // ...
  newline { true }
```
where `newline` is
```
  newline_char =
      [\r]     // Carriage Return U+000D
    | [\n]     // Line Feed U+000A
    | [\u0085] // Next Line U+0085
    | [\u000B] // Vertical Tab U+000B
    | [\u000C] // Form Feed U+000C
    | [\u2028] // Line Separator U+2028
    | [\u2029] // Paragraph Separator U+2029
  ;
  newline = "\r\n" | newline_char;
```

re2c 4.0.2

**Comment by skvadrik** [comment]
Thanks for reporting!

Here's a quick workaround until this is fixed properly:
```
re2c:YYPEEK = "getAsInt";
re2c:yych:literals = hex;
```
Here's the implementation of `getAsInt`:
```ocaml
let getAsInt str idx = Char.code (get str idx)
```
I tested on the following example (posting for completeness):
```ocaml
open String

type state = {
    yyinput: string;
    mutable yycursor: int;
    mutable yymarker: int;
}

let getAsInt str idx = Char.code (get str idx)

%{
    re2c:YYFN = ["lex;bool", "yyrecord;state"];
    re2c:YYPEEK = "getAsInt";
    re2c:yyfill:enable = 0;
    re2c:yych:literals = hex;

    newline_char =
        [\r]     // Carriage Return U+000D
      | [\n]     // Line Feed U+000A
      | [\u0085] // Next Line U+0085
      | [\u000B] // Vertical Tab U+000B
      | [\u000C] // Form Feed U+000C
      | [\u2028] // Line Separator U+2028
      | [\u2029] // Paragraph Separator U+2029
    ;
    newline = "\r\n" | newline_char;

    newline { true }
    *       { false }
%}

let main () =
    let st = {yyinput = "\n"; yycursor = 0; yymarker = 0}
    in if not (lex st) then raise (Failure "error")

let _ = main ()
```

The proper fix will likely require adding the information about valid escape sequences to the syntax file for each language. It will take a bit more time.

**Comment by eilvelia** [comment]
> Here's a quick workaround until this is fixed properly:

I use `sed -i "s/'\\\\f'/'\\\\x0C'/g" src/re2c_lexer.ml` as a workaround :)

**Comment by skvadrik** [comment]
Fixed in https://github.com/skvadrik/re2c/commit/54d0eaa4b56b83baa25934e3dec415da60257b5e.

**Comment by eilvelia** [comment]
I can confirm it doesn't generate invalid escapes anymore.

If you don't mind answering questions here, I have a bit of a problem with retrieving positions of a token when using YYFN. I.e., I perhaps need to add some code to the yyfn function to save the start position of a token or something similar. Is there a way to do something like that?

**Comment by skvadrik** [comment]
> If you don't mind answering questions here, I have a bit of a problem with retrieving positions of a token when using YYFN. I.e., I perhaps need to add some code to the yyfn function to save the start position of a token or something similar. Is there a way to do something like that?

There's no way at the moment to add entry code to `YYFN`, but I'm not sure it's needed in order to track token position. It usually can be done in semantic action or surrounding code, because token starts at the entry into the automaton and ends in semantic action (all the intermediate states represented by YYFN functions correspond to mid-token positions).

You can override YYPEEK or YYSKIP to do some additional bookkeeping, but I'm not sure it's needed. Can you explain a bit more what you are trying to do, maybe with a small example?

**Comment by eilvelia** [comment]
@skvadrik 

> or surrounding code

I presume that would've been easy without `--recursive-functions`, but with the option enabled the entry itself is kind of generated by re2c. I've thought about overriding YYPEEK, although that is too hacky.

Right now I did roughly:

```ocaml
%{local
  re2c:YYFN = ["main_lex;token", "yyrecord;yyrecord"];
  
  ws { main yyrecord } // main is called here, not main_lex
  newline { /* ... */ main yyrecord }
  // ...
  "token1" { Token1 }
  "token2" { Token2 }
%}

and main yyrecord =
  yyrecord.start_position <- yyrecord.yycursor;
  main_lex yyrecord
```

(I cannot place the "main" function above the local block because it won't make into the chain of mutually-recursive functions, but last generated function is conveniently generated with "and" so this works, even though the "and" declaration is invalid on its own.)

I guess it is good enough, even if not ideally convenient/intuitive.

Then it is used (simplifying) as follows:

```ocaml
let main_tokenizer yyrecord =
  let token = main yyrecord in
  (token, yyrecord.start_position, yyrecord.yycursor)
```

which is called by the parser.

I need positions around tokens for error messages, without taking into account skipped whitespace and things like that. (And without manually prepending every rule with a tag.)

For context, in OCaml the ocamlllex and sedlex lexer generators all work by saving the starting lexeme position automatically and then giving a (start_pos, end_pos) pair besides the token.

**Comment by skvadrik** [comment]
> I presume that would've been easy without --recursive-functions, but with the option enabled the entry itself is kind of generated by re2c.

Not really, even with `--goto-label` code model re2c does not allow prepending arbitrary code to each state. `YYFN` is a template function that gets generated for each state (as you've probably seen if you have looked at the generated code). In order to save token start and end you don't need code in *every* state, only in the start state and in the final state(s).

> I guess it is good enough, even if not ideally convenient/intuitive.

This looks good to me: you save position immediately before entering automaton. I would do the same.

> I need positions around tokens for error messages, without taking into account skipped whitespace and things like that. (And without manually prepending every rule with a tag.)

Tags are for saving positions in the middle of a rule. They are not just syntactic sugar - allowing the user to add some code that saves positions in an intermediate state may result in incorrect results (mid-rule actions are misleading: they may get executed unexpectedly for the user, because a DFA merges multiple NFA paths together, collapsing actions on different paths). So, if you need to save position in the middle of a rule, you should use a tag. But if you need to save positions only on entry/exit, you don't need tags (prepending each rule with a tag is certainly wasted effort).

> For context, in OCaml the ocamlllex and sedlex lexer generators all work by saving the starting lexeme position automatically and then giving a (start_pos, end_pos) pair besides the token.

Good point, this is on my TODO list.

**Comment by eilvelia** [comment]
> Not really, even with --goto-label code model re2c does not allow prepending arbitrary code to each state.

I was thinking it would allow me to more easily call the automaton recursively with custom code inbetween (without the need for additional function), e.g.:

```ocaml
let start_pos = ref 0
let rec lex yycursor =
  start_pos := yycursor; (* <-- prepended code *)
  (* ... generated goto ... *)
  %{
    re2c:yyfill:enable = 0;
    ws { lex yycursor }
    token { Token }
  %}
```

(I don't need to do it on _every_ state, yes. I was intending to override only one starting yyfn.)

But I'm guessing it's actually mostly the limitations of OCaml (of its mutually-recursive functions) that are relevant here. The way I did it only works because of implementation details of the generated code, which syntactically allow me to add another function to the mutually-recursive chain (unlike with the goto code model).

> But if you need to save positions only on entry/exit, you don't need tags (prepending each rule with a tag is certainly wasted effort).

Yes, sure.

Thank you for your answers, by the way :)

**Comment by eilvelia** [comment]
To clarify one misunderstanding:

> YYFN is a template function that gets generated for each state (as you've probably seen if you have looked at the generated code).

I was a bit confused. To be clear, earlier I was referring to the function name I specify in the YYFN parameter:

> `re2c:YYFN = ["lex;token"; "yyrecord;yyrecord"]`

I.e., I meant to override the "lex" function specifically, not all yyXXX functions generated based on its template.

**Comment by skvadrik** [comment]
> The way I did it only works because of implementation details of the generated code, which syntactically allow me to add another function to the mutually-recursive chain (unlike with the goto code model).

Relying on `and` is fine.  If you look at [this example](https://re2c.org/playground/?example=ocaml/fill/02_fill.re), it does the same as you do:
```
and lex_loop st count =
    st.token <- st.yycursor;
    try lex st count with Fill -> -1
```
So it won't get broken by accident.

By the way you can override the way code is generated (it's defined [here in the syntax file](https://github.com/skvadrik/re2c/blob/master/include/syntax/ocaml#L247)).

> I.e., I meant to override the "lex" function specifically, not all yyXXX functions generated based on its template.

Ah, I see. The idea was that you can always write a function that calls the start function and do whatever initialization you need there.

> Thank you for your answers, by the way :)

You are very welcome!

**Comment by eilvelia** [comment]
> If you look at this example, it does the same as you do:

Oh, nice, I didn't pay attention to that example!

Looking into my re2c lexers, there's actually `(![^]) { ... }` in some places because that's was the simplest thing to do (but not the most efficient), rather than create wrappers with initialization for every `%{local...%}` block with a distinct automata.

So I also have a bit of feature request ;). (Which should be similar to what you had in todo?) It would be quite convenient for me if there was `YYSTART` which does `yystart = yycursor;` on automata enter and is enabled by, for example, `re2c:yystart:enable = 1;`. With the goto code model, you can simply save yycursor just before the generated code, so this should be the most useful for recursive-functions.

**Comment by skvadrik** [comment]
> Looking into my re2c lexers, there's actually (![^]) { ... } in some places because that was the simplest thing to do (but not the most efficient), rather than create wrappers with initialization for every %{local...%} block with a distinct automaton.

It seems problematic for a few reasons: `[^]` consumes the symbol (I imagine you roll back `yycursor` in semantic action), the rule may be shadowed by another rule that matches a longer string, and the closing parenthesis may be unnecessary (unless you use it). `@start "" { ... }` or `@start / [^]` might be slightly better, but it still has the shadowing problem. A wrapper over the lex function seems much simpler to me.

> So I also have a bit of feature request ;). (Which should be similar to what you had in todo?) It would be quite convenient for me if there was YYSTART which does yystart = yycursor; on automata enter and is enabled by, for example, re2c:yystart:enable = 1;. With the goto code model, you can simply save yycursor just before the generated code, so this should be the most useful for recursive-functions.

Great, I've created a bug https://github.com/skvadrik/re2c/issues/517. I need to think over the interface. What you suggest is reasonable, but I also want to cover the end marker. In your case, do you need only the start marker or the end one as well?

**Comment by eilvelia** [comment]
> It seems problematic for a few reasons: [^] consumes the symbol

Ah, in this case I want to consume the symbol. To be clear, `(![^])` is the whole rule, I wanted to highlight there were groups spanning over the whole rule because that was simplest (also reuses existing code I have to work with `yytl0..yytr0`). Shadowing by other rules was intentional.

Well, I now write
```ocaml
and <fun> yyrecord <other-args> =
  yyrecord.yystart <- yyrecord.yycursor;
  <fun>_body yyrecord <other-args>
```
after every `%{local` block with yyfn (the name in the yyfn parameter is `<fun>_body`).

> What you suggest is reasonable, but I also want to cover the end marker. In your case, do you need only the start marker or the end one as well?

I think I only need the start marker because the end one in a semantic action is `yycursor`? (As an example, ocamllex stores `start` and `current` positions.)

**Comment by skvadrik** [comment]
> I think I only need the start marker because the end one in a semantic action is yycursor? (As an example, ocamllex stores start and current positions.)

Ah, right. Makes sense.

**Comment by skvadrik** [comment]
I'm thinking of making YYSTART more generic, to let the user add arbitrary code at the lexer entry. By default it will be empty, meaning that no code is generated. You can define it as `re2c:YYSTART = "yystart = yycursor;";` to save token start. Then we can omit the boolean configuration altogether -- old re2c versions will complain about nonexistent configuration. What do you think?

There's a question what exactly should be considered "lexer entry" in the case of conditions and/or storable states (it can be the `yy0` state or condiiton/state dispatch).

**Comment by eilvelia** [comment]
> By default it will be empty, meaning that no code is generated. You can define it as re2c:YYSTART = "yystart = yycursor;"; to save token start. Then we can omit the boolean configuration altogether -- old re2c versions will complain about nonexistent configuration. What do you think?

Looks good, I think. I was thinking of an easy-to-use solution in which `re2c:YYSTART:enable = 1;` works regardless of yyrecord/language. But I guess generality is quite useful here. In my case, I also count newlines, and in the main lexer, save the line count + line offset besides yycursor, but most of my `local` lexers need yystart (saved yycursor) only. I'm not sure about the name of it, other options can be `YYPRELUDE`/`YYPROLOGUE`/`YYBEGIN` (if the intended goal is general initialization, not just saving yycursor). I think with conditionals, I imagined it would be executed at the very start, i.e. same as if you called the lexer from a wrapper with custom initialization. (BTW, I've sent you a message on irc (libera), if you haven't seen it.)

**Comment by skvadrik** [comment]
> I think with conditionals, I imagined it would be executed at the very start, i.e. same as if you called the lexer from a wrapper with custom initialization.

It may be better to generate the initialization code at the entry to the sub-lexer for each condition, since the user may want to jump to some condition directly, bypassing the dispatch. The way it is done in this example: http://re2c.org/playground/?example=ocaml/conditions/parse_u32_conditions.re (`yyfnbin` etc. functions are called directly, not via `parse`).

**Comment by eilvelia** [comment]
> since the user may want to jump to some condition directly, bypassing the dispatch. The way it is done in this example: http://re2c.org/playground/?example=ocaml/conditions/parse_u32_conditions.re (yyfnbin etc. functions are called directly, not via parse).

Hmm, you are right. _(Changing my code to use `yyfn*`.)_ It would make sense then to duplicate the initialization into the initial state of every condition. In that case, such a configuration would be even more useful.

**Comment by skvadrik** [comment]
I've been thinking about YYSTART, and it does not really behave like an API - it's more like a semantic action, only one without a rule that gets executed at lexer entry (similar to zero condition `<>`). So it would make more sense to let it syntactically look like semantic action, so that the user can write an arbitrary block of code there, without having to squeeze it into a string. I've thought of using `^` for the start rule (in part because it's the start of string symbol in POSIX RE), like this:
```
^ {
    // any code ...
}
```

Inheritance of rules between blocks works the same as with configurations.

It's a slightly bigger change than YYSTART (adding a new type of rule), but it makes more sense to me. @eilvelia What do you think? @helly25 any opinions? You may have to read a few comments up the thread from [here](https://github.com/skvadrik/re2c/issues/516#issuecomment-2561497585) to get the context.

**Comment by eilvelia** [comment]
> I've been thinking about YYSTART, and it does not really behave like an API - it's more like a semantic action, only one without a rule that gets executed at lexer entry (similar to zero condition <>). So it would make more sense to let it syntactically look like semantic action, so that the user can write an arbitrary block of code there, without having to squeeze it into a string.

Oh, yes, I like it. One thing to note is that it wouldn't really be like other semantic actions, that is, it is not executed in a final state, and `^` means "start of automaton" and not "start of input" (dissimilar to `$`). Although the `^` symbol may still be the best option.

With `rules`, should `^` also be an [exception](https://re2c.org/manual/manual_c.html#reusable-blocks) to overriding like `*`, `$`, and `<!>`?

(It's a bit unfortunate that this thread is below an unrelated issue. :P)

**Comment by skvadrik** [comment]
> One thing to note is that it wouldn't really be like other semantic actions, that is, it is not executed in a final state, and ^ means "start of automaton" and not "start of input" (dissimilar to $).

True. But with re2c `^` will never have the same meaning as in POSIX RE because matching always starts from the beginning, so we won't end up in a situation when we need `^` for "start of string", but it's already taken.

> With rules, should ^ also be an [exception](https://re2c.org/manual/manual_c.html#reusable-blocks) to overriding like *, $, and <!>?

Yes, exactly.

> (It's a bit unfortunate that this thread is below an unrelated issue. :P)

Yeah, I should have moved it to a new issue, but it's probably too late now.

**Comment by skvadrik** [comment]
Let's discuss here: https://github.com/skvadrik/re2c/issues/521, as I thought of a few more potential improvements.
